--- ORBit2-2.14.17/linc2/src/linc-connection.c.eagain	2008-08-31 10:04:44.000000000 -0400
+++ ORBit2-2.14.17/linc2/src/linc-connection.c	2012-10-15 08:28:06.085674961 -0400
@@ -16,6 +16,7 @@
 #include <string.h>
 #include <stdio.h>
 #include <ctype.h>
+#include <glib/gtimer.h> /* for g_usleep */
 
 #ifdef LINK_SSL_SUPPORT
 #  include <openssl/ssl.h>
@@ -45,6 +46,10 @@ static GList *cnx_list = NULL;
 #define CNX_AND_LIST_UNLOCK(cnx) CNX_UNLOCK(cnx); /* for now */
 #define CNX_IS_LOCKED(cnx) link_is_locked()
 
+#ifndef WAIT_RETRY
+#  define WAIT_RETRY    10000 /* microseconds */
+#endif
+
 static gboolean link_connection_io_handler (GIOChannel  *gioc,
 					    GIOCondition condition,
 					    gpointer     data);
@@ -652,7 +657,13 @@ link_connection_do_initiate (LinkConnect
 	}
 #endif
 
+retry:
 	LINK_TEMP_FAILURE_RETRY_SOCKET (connect (fd, saddr, saddr_len), rv);
+	if (rv == -1L && errno == EAGAIN) {
+		g_usleep (WAIT_RETRY);
+		goto retry;
+	}
+
 #ifdef HAVE_WINSOCK2_H
 	if (rv == SOCKET_ERROR) {
 		if ((options & LINK_CONNECTION_NONBLOCKING) &&
